architecture_records/002-acceleration-framework.md

```diff
@@ -0,0 +1,480 @@
+# Training Enhancements Framework
+
+**Deciders(s)**: Sukriti Sharma (sukriti.sharma4@ibm.com), Raghu Ganti (rganti@us.ibm.com), Laura Wynter (lwynter@sg.ibm.com), Fabian Lim (flim@sg.ibm.com), Aaron Chew (aaron.chew1@ibm.com)
+**Date (YYYY-MM-DD)**:  2024-04-11
+**Obsoletes ADRs**:  NA
+**Modified By ADRs**:  NA
+**Relevant Issues**: [116](https://github.com/foundation-model-stack/fms-hf-tuning/pull/116)
+
+- [Summary and Objective](#summary-and-objective)
+  - [Motivation](#motivation)
+  - [User Benefit](#user-benefit)
+- [Decision](#decision)
+  - [Alternatives Considered](#alternatives-considered)
+- [Consequences](#consequences)
+- [Detailed Design](#detailed-design)
+
+## Summary and Objective
+
+Design and implement a framework to include custom acceleration tools into [`sft_trainer.py`], that improve training metrics such as GPU memory consumption, training speed, etc.
+
+<!--
+Context goes here.
+
+Describe the forces at play, including technological, political, social, and project local. These forces are likely in tension, and should be called out as such. The language in this section is value-neutral. It is simply describing facts.
+-->
+
+### Motivation
+
+Currently `sft_trainer.py` only can access those tools already integrated in HF. Due to rapid developments in AI training methodologies, these technologies are quickly considered "standard", for example:
+1. LoRA adapters from [PEFT](https://github.com/huggingface/peft).
+2. Prefix tuning from [PEFT](https://github.com/huggingface/peft).
+3. FSDP training from [accelerate](https://github.com/huggingface/accelerate).
+
+Below are various reasons for a framework to integrate custom training tools into `sft_trainer.py`. 
+* Enable quick integrations of open-source techniques that have yet to be integrated into Huggingface.
+* Enable integrations of custom techniques developed by IBM researchers, that are not planned be integrated into Huggingface.
```

Comment:
I think more specific example use-cases might help clarify the motivation here, or at least the nature of those use cases. What are some of the open source techniques this could be used to implement?

And conversely what are examples of use-cases that would not be supported?

I think anything that modifies the logic of the training loop would not be supported (In the docs for Callbacks it says "For customizations that require changes in the training loop, you should subclass [Trainer](https://huggingface.co/docs/transformers/v4.40.1/en/main_classes/trainer#transformers.Trainer) and override the methods you need")

... and so I think the idea is this is all about modifying just the model, and then callbacks would just be for alternate checkpointing mechanisms?

Reply:
Following up on the previous comment, those 3 examples i have listed are quite canonical. Basically these would be use cases that 
- need to load the model differently (i.e., cannot use AutoModel.from_pretrained), like loading a quantized checkpoint, or some custom sharding strategy no supported OOTB in FSDP like in the megablocks use case.
- need to replace layers in ways not supported (e.g., like when there is a new triton kernel release not yet integrated into HF).
- need to modify some FSDP configuration. This is needed in megablocks, and [there is already one example of such thing being done currently in sft_trainer.py](https://github.com/fabianlim/fms-hf-tuning/blob/2c19b708accc5704c01c01efe0f62d1a4735752a/tuning/sft_trainer.py#L277-L280). 

> ... and so I think the idea is this is all about modifying just the model, and then callbacks would just be for alternate checkpointing mechanisms? Does model loading happen outside of the trainer class?

Callbacks may not be always used for checkpointing. If Im thinking ahead of the megablocks use case, then I plan to use `TrainerCallback.on_init_end` to modify the configuration of the `fsdp_plugin`, similar to [what is done here](https://github.com/fabianlim/fms-hf-tuning/blob/2c19b708accc5704c01c01efe0f62d1a4735752a/tuning/sft_trainer.py#L277-L280).